---
title: "Class 13: RNA Seq Analysis Mini Project"
author: "Courtney Anderson (PID:A69038035)"
format: pdf
---
## Background
Today we will eun through a complete RNA Seq Analyisis:

The data for for hands-on session comes from GEO entry: GSE37704, which is associated with the following publication:

Trapnell C, Hendrickson DG, Sauvageau M, Goff L et al. "Differential analysis of gene regulation at transcript resolution with RNA-seq". Nat Biotechnol 2013 Jan;31(1):46-53. PMID: 23222703
The authors report on differential analysis of lung fibroblasts in response to loss of the developmental transcription factor HOXA1. Their results and others indicate that HOXA1 is required for lung fibroblast and HeLa cell cycle progression. In particular their analysis show that "loss of HOXA1 results in significant expression level changes in thousands of individual transcripts, along with isoform switching events in key regulators of the cell cycle". For our session we have used their Sailfish gene-level estimated counts and hence are restricted to protein-coding genes only.

## Section 1: Differential Expression Analysis

Load in data 
```{r, message=F}
library(DESeq2)
metaFile <- "/Users/courtneyanderson/Desktop/R_Files/BGGN213/Class 13/GSE37704_metadata.csv"
countFile <- "/Users/courtneyanderson/Desktop/R_Files/BGGN213/Class 13/GSE37704_featurecounts.csv"
```

```{r}
colData = read.csv(metaFile, row.names=1)
head(colData)
```
```{r}
countData = read.csv(countFile, row.names=1)
head(countData)
```

> Q1. Complete the code below to remove the troublesome first column from countData: 

```{r}
countData <- as.matrix(countData[,-1])
head(countData)
```
>Q2. Complete the code below to filter countData to exclude genes (i.e. rows) where we have 0 read count across all samples (i.e. columns).Tip: What will rowSums() of countData return and how could you use it in this context?

```{r}
countData <- countData[rowSums(countData) > 0, ]
head(countData)
```

## Running DESeq2
```{r,message=F}
dds = DESeqDataSetFromMatrix(countData=countData,
                             colData=colData,
                             design=~condition)
dds = DESeq(dds)
dds
```
```{r}
res = results(dds)
```

>Q3. Call the summary() function on your results to get a sense of how many genes are up or down-regulated at the default 0.1 p-value cutoff.

```{r}
summary(res)
```
## Volcano plot 
```{r}
plot( res$log2FoldChange, -log(res$padj) )
```
```{r}
library(ggplot2)

ggplot(res)+
  aes(res$log2FoldChange, -log(res$padj)) +
  geom_point() +
  geom_vline(xintercept = c(-2,2), col="red") +
  geom_hline(yintercept = -log(0.05), col="red")
```

>Q4. Improve this plot by completing the below code, which adds color and axis labels

```{r}
# Make a color vector for all genes
mycols <- rep("gray", nrow(res))

# Color red the genes with absolute fold change above 2
mycols[ abs(res$log2FoldChange) > 2 ] <- "red"

# Color blue those with adjusted p-value less than 0.01
# and absolute fold change more than 2
inds <- (res$padj < 0.01) & (abs(res$log2FoldChange) > 2)
mycols[ inds ] <- "blue"

# Plot
plot(res$log2FoldChange, -log(res$padj), col=mycols,
     xlab="Log2(FoldChange)", ylab="-Log(P-value)")

```
## Adding Gene Annotation 

>Q5. Use the mapIDs() function multiple times to add SYMBOL, ENTREZID and GENENAME annotation to our results by completing the code below.

```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")

columns(org.Hs.eg.db)

res$symbol = mapIds(org.Hs.eg.db,
                    keys=row.names(res), 
                    keytype="ENSEMBL",
                    column="SYMBOL",
                    multiVals="first")

res$entrez = mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="ENTREZID",
                    multiVals="first")

res$name =   mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="GENENAME",
                    multiVals="first")
head(res, 10)
```
```

>Q6. Finally for this section let's reorder these results by adjusted p-value and save them to a CSV file in your current project directory.

```{r}
res <- res[order(res$pvalue), ]
write.csv(res, file = "deseq_results.csv")
```


## Section 2: Pathway Analysis

```{r,message=F}
library(pathview)
```

```{r}
library(gage)
library(gageData)

data(kegg.sets.hs)
data(sigmet.idx.hs)

# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

# Examine the first 3 pathways
head(kegg.sets.hs, 3)
## $`hsa00232 Caffeine metabolism`
```

```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```
```{r}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)
attributes(keggres)
```
```{r}
# Look at the first few down (less) pathways
head(keggres$less)
```
```{r, message=F}
pathview(gene.data=foldchanges, pathway.id="hsa04110")
```
```{r}
library(png)
library(grid)

# Read the image file (replace 'my_image.png' with your filename)
img <- readPNG("hsa04110.pathview.png")

# Display the image
grid.raster(img)
```

```{r}
## Focus on top 5 upregulated pathways here for demo purposes only
keggrespathways <- rownames(keggres$greater)[1:5]

# Extract the 8 character long IDs part of each string
keggresids = substr(keggrespathways, start=1, stop=8)
keggresids
```

```{r}
pathview(gene.data=foldchanges, pathway.id=keggresids, species="hsa")
```

```{r}
library(png)
library(grid)

# Read the image file (replace 'my_image.png' with your filename)
img2 <- readPNG("hsa04142.pathview.png")
# Display the image
grid.raster(img2)
```

```{r}
img3 <- readPNG("hsa04330.pathview.png")
grid.raster(img3)
```

```{r}
img4 <- readPNG("hsa04630.pathview.png")
grid.raster(img4)
```

```{r}
img5 <- readPNG("hsa04640.pathview.png")
grid.raster(img5)
```

## Section 3: Gene Ontology 

```{r}
data(go.sets.hs)
data(go.subs.hs)

# Focus on Biological Process subset of GO
gobpsets = go.sets.hs[go.subs.hs$BP]

gobpres = gage(foldchanges, gsets=gobpsets, same.dir=TRUE)

lapply(gobpres, head)
```
## Section 4: Reactome Analysis

```{r}
sig_genes <- res[res$padj <= 0.05 & !is.na(res$padj), "symbol"]
print(paste("Total number of significant genes:", length(sig_genes)))
```
```{r}
write.table(sig_genes, file="significant_genes.txt", row.names=FALSE, col.names=FALSE, quote=FALSE)
```

>Q7. What pathway has the most significant “Entities p-value”? Do the most significant pathways listed match your previous KEGG results? What factors could cause differences between the two methods? The most significant pathways are listed below, however, they are cell cycle, mitotic prometaphase, mitotic spindle checkpoint, etc.I would say that my KEGG results match this analysis since the top hit is cell cycle, but I felt like the listed pathways were more specific than the KEGG results. The reactome analysis would be more specifc since it is divided into sub pathways whereas KEGG groupd genes into broader functional maps.

```{r}
library(png)
library(grid)

# Read the image file (replace 'my_image.png' with your filename)
img <- readPNG("Siggeneanalysis.png")

# Display the image
grid.raster(img)
```

